small/lsregion_asan.c

```diff
@@ -30,17 +30,56 @@
  */
 #include "lsregion.h"
 
+SMALL_NO_SANITIZE_ADDRESS static struct lsregion_allocation *
+spare_alloc(struct lsregion *lsregion)
+{
+	if (!rlist_empty(&lsregion->allocations)) {
+		struct lsregion_allocation *alloc = NULL;
+		alloc = rlist_last_entry(&lsregion->allocations, typeof(*alloc),
+					 link);
+		if (alloc->id == LSLAB_NOT_USED_ID)
+			return alloc;
+	}
+	return NULL;
+}
+
+SMALL_NO_SANITIZE_ADDRESS static struct lsregion_allocation *
+alloc_reserve(struct lsregion *lsregion, size_t size, size_t alignment)
+{
+	struct lsregion_allocation *alloc = small_asan_alloc(size, alignment,
+							     sizeof(*alloc));
+	alloc->size = size;
+	alloc->used = 0;
+	alloc->id = LSLAB_NOT_USED_ID;
+	alloc->alignment = alignment;
+	rlist_add_tail_entry(&lsregion->allocations, alloc, link);
+	return alloc;
+}
+
+void *
+lsregion_aligned_reserve(struct lsregion *lsregion, size_t size,
+			 size_t alignment)
+{
+	/* Double reserve is prohibited. */
+	small_asan_assert(spare_alloc(lsregion) == NULL);
+	struct lsregion_allocation *alloc = alloc_reserve(lsregion, size,
+							  alignment);
+	return small_asan_payload_from_header(alloc);
+}
+
 SMALL_NO_SANITIZE_ADDRESS void *
 lsregion_aligned_alloc(struct lsregion *lsregion, size_t size, size_t alignment,
 		       int64_t id)
 {
-	struct lsregion_allocation *alloc =
-			small_asan_alloc(size, alignment,
-					 sizeof(struct lsregion_allocation));
-	alloc->size = size;
+	struct lsregion_allocation *alloc = spare_alloc(lsregion);
+	if (alloc == NULL)
+		alloc = alloc_reserve(lsregion, size, alignment);
+	small_asan_assert(alloc->size >= size && alloc->alignment == alignment);
+	alloc->used = size;
 	alloc->id = id;
-	rlist_add_tail_entry(&lsregion->allocations, alloc, link);
 	lsregion->used += size;
+	char *payload = small_asan_payload_from_header(alloc);
+	ASAN_POISON_MEMORY_REGION(payload + size, alloc->size - size);
```

Comment:
We tend to have prefixes in other implementation. Also I'd suggest a bit different names:

alloc_reserve -> lsregion_prepare_buf (consistency with other implementation).
spare_alloc - > lsregion_reserved_alloc.

Reply:
Thanks! Done:
```diff
diff --git a/small/lsregion_asan.c b/small/lsregion_asan.c
index 24fece5..036742f 100644
--- a/small/lsregion_asan.c
+++ b/small/lsregion_asan.c
@@ -44,7 +44,7 @@ spare_alloc(struct lsregion *lsregion)
 }
 
 static SMALL_NO_SANITIZE_ADDRESS struct lsregion_allocation *
-alloc_reserve(struct lsregion *lsregion, size_t size, size_t alignment)
+lsregion_prepare_buf(struct lsregion *lsregion, size_t size, size_t alignment)
 {
 	struct lsregion_allocation *alloc = small_asan_alloc(size, alignment,
 							     sizeof(*alloc));
@@ -62,8 +62,8 @@ lsregion_aligned_reserve(struct lsregion *lsregion, size_t size,
 {
 	/* Double reserve is prohibited. */
 	small_asan_assert(spare_alloc(lsregion) == NULL);
-	struct lsregion_allocation *alloc = alloc_reserve(lsregion, size,
-							  alignment);
+	struct lsregion_allocation *alloc = lsregion_prepare_buf(lsregion, size,
+								 alignment);
 	return small_asan_payload_from_header(alloc);
 }
 
@@ -73,7 +73,7 @@ lsregion_aligned_alloc(struct lsregion *lsregion, size_t size, size_t alignment,
 {
 	struct lsregion_allocation *alloc = spare_alloc(lsregion);
 	if (alloc == NULL)
-		alloc = alloc_reserve(lsregion, size, alignment);
+		alloc = lsregion_prepare_buf(lsregion, size, alignment);
 	small_asan_assert(alloc->size >= size && alloc->alignment == alignment);
 	alloc->used = size;
 	alloc->id = id;
```
```diff
diff --git a/small/lsregion_asan.c b/small/lsregion_asan.c
index 036742f..c153407 100644
--- a/small/lsregion_asan.c
+++ b/small/lsregion_asan.c
@@ -31,7 +31,7 @@
 #include "lsregion.h"
 
 static SMALL_NO_SANITIZE_ADDRESS struct lsregion_allocation *
-spare_alloc(struct lsregion *lsregion)
+lsregion_reserved_alloc(struct lsregion *lsregion)
 {
 	if (!rlist_empty(&lsregion->allocations)) {
 		struct lsregion_allocation *alloc = NULL;
@@ -61,7 +61,7 @@ lsregion_aligned_reserve(struct lsregion *lsregion, size_t size,
 			 size_t alignment)
 {
 	/* Double reserve is prohibited. */
-	small_asan_assert(spare_alloc(lsregion) == NULL);
+	small_asan_assert(lsregion_reserved_alloc(lsregion) == NULL);
 	struct lsregion_allocation *alloc = lsregion_prepare_buf(lsregion, size,
 								 alignment);
 	return small_asan_payload_from_header(alloc);
@@ -71,7 +71,7 @@ SMALL_NO_SANITIZE_ADDRESS void *
 lsregion_aligned_alloc(struct lsregion *lsregion, size_t size, size_t alignment,
 		       int64_t id)
 {
-	struct lsregion_allocation *alloc = spare_alloc(lsregion);
+	struct lsregion_allocation *alloc = lsregion_reserved_alloc(lsregion);
 	if (alloc == NULL)
 		alloc = lsregion_prepare_buf(lsregion, size, alignment);
 	small_asan_assert(alloc->size >= size && alloc->alignment == alignment);
```

Resolving the comment since the PR is approved.